sdk/core/azure-core/src/http/websockets/websockets_impl.cpp

```diff
@@ -0,0 +1,884 @@
+    auto curlWebSockets
+        = std::make_shared<Azure::Core::Http::WebSockets::CurlWebSocketTransport>(transportOptions);
+
+    m_transport = std::static_pointer_cast<WebSocketTransport>(curlWebSockets);
+    m_options.Transport.Transport = std::static_pointer_cast<HttpTransport>(curlWebSockets);
+#endif
+
+    std::vector<std::unique_ptr<Azure::Core::Http::Policies::HttpPolicy>> perCallPolicies{};
+    std::vector<std::unique_ptr<Azure::Core::Http::Policies::HttpPolicy>> perRetryPolicies{};
+    // If the caller has told us a service name, add the telemetry policy to the pipeline to add
+    // a user agent header to the request.
+    if (!m_options.ServiceName.empty())
+    {
+      perCallPolicies.push_back(
+          std::make_unique<Azure::Core::Http::Policies::_internal::TelemetryPolicy>(
+              m_options.ServiceName, m_options.ServiceVersion, m_options.Telemetry));
+    }
+    Azure::Core::Http::_internal::HttpPipeline openPipeline(
+        m_options, std::move(perRetryPolicies), std::move(perCallPolicies));
+
+    Azure::Core::Http::Request openSocketRequest(
+        Azure::Core::Http::HttpMethod::Get, m_remoteUrl, false);
+
+    // Generate the random request key. Only used when the transport doesn't support websockets
+    // natively.
+    auto randomKey = GenerateRandomKey();
+    auto encodedKey = Azure::Core::Convert::Base64Encode(randomKey);
+    if (!m_transport->HasBuiltInWebSocketSupport())
+    {
+      // If the transport doesn't support WebSockets natively, set the standardized WebSocket
+      // upgrade headers.
+      openSocketRequest.SetHeader("Upgrade", "websocket");
+      openSocketRequest.SetHeader("Connection", "upgrade");
+      openSocketRequest.SetHeader("Sec-WebSocket-Version", "13");
+      openSocketRequest.SetHeader("Sec-WebSocket-Key", encodedKey);
+    }
+    if (!m_options.Protocols.empty())
+    {
+
+      std::string protocols;
+      for (auto const& protocol : m_options.Protocols)
+      {
+        protocols += protocol;
+        protocols += ", ";
+      }
+      protocols = protocols.substr(0, protocols.size() - 2);
+      openSocketRequest.SetHeader("Sec-WebSocket-Protocol", protocols);
+    }
+    for (auto const& additionalHeader : m_headers)
+    {
+      openSocketRequest.SetHeader(additionalHeader.first, additionalHeader.second);
+    }
+    std::string remoteOrigin;
+    remoteOrigin = m_remoteUrl.GetScheme();
+    remoteOrigin += "://";
+    remoteOrigin += m_remoteUrl.GetHost();
+    openSocketRequest.SetHeader("Origin", remoteOrigin);
+
+    // Send the connect request to the WebSocket server.
+    auto response = openPipeline.Send(openSocketRequest, context);
+
+    // Ensure that the server thinks we're switching protocols. If it doesn't,
+    // fail immediately.
+    if (response->GetStatusCode() != Azure::Core::Http::HttpStatusCode::SwitchingProtocols)
+    {
+      throw Azure::Core::Http::TransportException("Unexpected handshake response");
+    }
```

Comment:
```suggestion
    auto statusCode = response->GetStatusCode();
    if (statusCode != Azure::Core::Http::HttpStatusCode::SwitchingProtocols)
    {
      throw Azure::Core::Http::TransportException("Unexpected handshake response:"+statusCode);
    }
```

Reply:
`TransportException` already includes the status code in the text:

```
HTTP Transport error: Error while sending a request. Error Code: 12029: A connection with the server could not be established
```